README
NAME
    File::SafeDO -- safer do file for perl

SYNOPSIS
      use File::SafeDO qw(
            DO
            doINCLUDE
      );

      $rv = DO($file,[optional no warnings string])

DESCRIPTION
    * $rv = DO($file,[optional] "no warnings string");
        This is a fancy 'do file'. It first checks that the file exists and
        is readable, then does a 'do file' to pull the variables and
        subroutines into the current name space. The 'do' is executed with
        full perl warnings so that syntax and construct errors are reported
        to STDERR. A string of no warnings may optionally be specified as a
        second argument. This is equivalent to saying:

          no warnings qw(string of no values);

        See: man perllexwarnings for a full listing of warning names.

          input:        file/path/name,
                    [optional] string of "no" warnings
          returns:      last value in file
                    or  undef on error
                    prints warning

          i.e. DO('myfile','once redefine');

        This will execute 'myfile' safely and suppress 'once' and 'redefine'
        warnings to STDERR.

    * $rv = doINCLUDE($file,[optional] "no warnings string");
        The function is similar to DO above with the addition of recursive
        loads.

        Function will recursively load a file which returns a hash pointer
        with the a key of the form:

                'INCLUDE' => somefile. 

        The file which it loads may contain only HASHs or SUBs. The HASH
        KEYS will be promoted into the parent hash, augmenting and replacing
        existing keys already present. Subroutines are simply imported into
        the name space as is the case with a 'do' or 'require'.

DEPENDENCIES
            none

EXPORT_OK
            DO
            doINCLUDE

AUTHOR
    Michael Robinton, michael@bizsystems.com

COPYRIGHT
    Copyright 2003 - 2008, Michael Robinton & BizSystems This program is
    free software; you can redistribute it and/or modify it under the terms
    of the GNU General Public License as published by the Free Software
    Foundation; either version 2 of the License, or (at your option) any
    later version.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
    Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.



Changes
Revision history for Perl extension File::SafeDO

0.13 Mon Jan 27 11:59:58 PST 2014
        modify Makefile.PL to bypass missing 'pod2text'

0.12  Sun Nov 16 15:37:50 PST 2008
	Add the capability to INCLUDE additional file(s)
	within a returned HASH object with the "INCLUDE"
	and the doINCLUDE function.

0.11  Tue Jul  5 19:05:39 PDT 2005
	Add the capability to suppress certain warnings during
	the DO process with an optional command switch.

0.10  Mon Mar 14 20:48:03 PST 2005
	Revision number 0.06 was and OOPS! number wise.
	The revision level in the file was actually 0.03
	sigh.... so I've upped it to 0.10

	The "do $file" function in the previous revision
	called the 'do' function in the scope of THIS package
	when what we really want is the scope of the 'caller' 
	package. This release corrects this defect with an
	eval that executes a "do $file" in the scope of the
	caller by inserting an anonymous subroutine into
	the calling package.
	
was really 0.03
0.06  Tue Mar  12 12:42:10 PST 2004
	initial release

	adapteded from Net::DNSBL::Utilities v0.02


Makefile.PL
use ExtUtils::MakeMaker;
use Config;

my $pkg = 'File::SafeDO';
$pkg =~ /[^:]+$/;
my $module = $& .'.pm';

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

WriteMakefile(
    'NAME'		=> $pkg,
    'VERSION_FROM'	=> $module, # finds $VERSION
    'clean'		=> { FILES    => "*~ tmp*"},
    'dist'		=> {COMPRESS=>'gzip', SUFFIX=>'gz'}
);

sub MY::top_targets {
  package MY;
  my $inherited = shift->SUPER::top_targets(@_);
  $inherited =~ s/(pure_all\s+::.+)/$1 README/;
  $inherited;
}

sub MY::post_constants {
  my $post_constants = $Config{scriptdirexp} .'/pod2text';
  if (-e $post_constants) {
  my  $post_constants = q| 
MY_POD2TEXT = |. $post_constants .q|
|;
  } else {
    $post_constants = q|
MY_POD2TEXT = echo "perl distro missing 'pod2text' to convert " |;
  }
}  

sub MY::postamble {
  package MY;
  my $postamble = q|
README  : |. $module .q|
	@$(MY_POD2TEXT) |. $module .q| > README

|;
}


SafeDO.pm
#!/usr/bin/perl

package File::SafeDO;
use strict;
#use diagnostics;

use vars qw($VERSION @ISA @EXPORT_OK);
require Exporter;
@ISA = qw(Exporter);

$VERSION = do { my @r = (q$Revision: 0.13 $ =~ /\d+/g); sprintf "%d."."%02d" x $#r, @r };

@EXPORT_OK = qw(
        DO
	doINCLUDE
);

use Config;

=head1 NAME

File::SafeDO -- safer do file for perl

=head1 SYNOPSIS

  use File::SafeDO qw(
        DO
	doINCLUDE
  );

  $rv = DO($file,[optional no warnings string])

=head1 DESCRIPTION

=over 4

=item * $rv = DO($file,[optional] "no warnings string");

This is a fancy 'do file'. It first checks that the file exists and is
readable, then does a 'do file' to pull the variables and subroutines into
the current name space. The 'do' is executed with full perl warnings so that 
syntax and construct errors are reported to STDERR. A string of B<no
warnings> may optionally be specified as a second argument. This is
equivalent to saying:

  no warnings qw(string of no values);

See: man perllexwarnings for a full listing of warning names.

  input:	file/path/name,
	    [optional] string of "no" warnings
  returns:	last value in file
	    or	undef on error
	    prints warning

  i.e. DO('myfile','once redefine');

This will execute 'myfile' safely and suppress 'once' and 'redefine'
warnings to STDERR.

=cut

sub xDO($;$) {
  my($file,$nowarnings) = @_;
  return undef unless
	$file &&
	-e $file &&
	-f $file &&
	-r $file;
  $_ = $Config{perlpath};		# bring perl into scope
  if ($nowarnings) {
    return undef if eval q|system($_, '-Mwarnings', "-M-warnings qw($nowarnings)", $file)|;
  } else {
    return undef if eval q|system($_, '-w', $file)|;
  }
# poke anonymous subroutine into calling package so vars and subs will import
  my $caller = caller;
# execute 'do $file;' in calling package
   &{eval "package $caller; sub { my \$file = shift; do \$file;};";}($file);
}

sub DO($;$) {
  my($file,$nowarnings) = @_;
  my $caller = caller;
  @_ = ($file,$nowarnings,$caller,0);
  goto &_doFILE;
}

=item * $rv = doINCLUDE($file,[optional] "no warnings string");

The function is similar to B<DO> above with the addition of recursive loads.

Function will recursively load a file which returns a hash pointer with the
a key of the form:
 
	'INCLUDE' => somefile. 

The file which it loads may contain only HASHs or SUBs. The HASH KEYS will
be promoted into the parent hash, augmenting and replacing existing keys
already present. Subroutines are simply imported into the name
space as is the case with a 'do' or 'require'.

=back

=cut

sub doINCLUDE($;$) {
  my($file,$nowarnings) = @_;
  my $caller = caller;
  @_ = ($file,$nowarnings,$caller,1);
  goto &_doFILE;
}

sub _doFILE($$$$) {
  my($file,$nowarnings,$caller,$recurs) = @_;
  return undef unless
	$file &&
	-e $file &&
	-f $file &&
	-r $file;
  $_ = $Config{perlpath};		# bring perl into scope
  if ($nowarnings) {
    return undef if eval q|system($_, '-Mwarnings', "-M-warnings qw($nowarnings)", $file)|;
  } else {
    return undef if eval q|system($_, '-w', $file)|;
  }
# poke anonymous subroutine into calling package so vars and subs will import
# execute 'do $file;' in calling package
  my $rv = &{eval "package $caller; sub { my \$file = shift; do \$file;};";}($file);
  return $rv unless $recurs && 
	UNIVERSAL::isa($rv,'HASH') &&
	exists $rv->{INCLUDE};
  my $rrv = &_doFILE($rv->{INCLUDE},$nowarnings,$caller,1);
  return $rv unless $rrv &&
	UNIVERSAL::isa($rv,'HASH');
  my @keys = keys %{$rrv};
  @{$rv}{@keys} = @{$rrv}{@keys};
  return $rv;
}

=head1 DEPENDENCIES

	none

=head1 EXPORT_OK

	DO
	doINCLUDE

=head1 AUTHOR

Michael Robinton, michael@bizsystems.com

=head1 COPYRIGHT

Copyright 2003 - 2008, Michael Robinton & BizSystems
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or 
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of 
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the  
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

=cut

1;
